File: Yona/Yona/Models/Attachment.swift
//
//  Attachment.swift
//  Yona
//
//  A document attached to a tile (the file lives in Supabase Storage; this is
//  its row in the `attachments` table).
//

import Foundation

struct Attachment: Identifiable, Decodable, Hashable {
    let id: UUID
    let tileId: UUID
    let filename: String
    let storagePath: String
    let contentType: String?
    let sizeBytes: Int?
    let createdAt: Date

    enum CodingKeys: String, CodingKey {
        case id, filename
        case tileId = "tile_id"
        case storagePath = "storage_path"
        case contentType = "content_type"
        case sizeBytes = "size_bytes"
        case createdAt = "created_at"
    }

    /// Human-readable size, e.g. "1.2 MB"; nil if unknown.
    var displaySize: String? {
        guard let sizeBytes else { return nil }
        return ByteCountFormatter.string(fromByteCount: Int64(sizeBytes), countStyle: .file)
    }
}

File: android/app/src/main/java/com/yona/app/core/Attachment.kt
package com.yona.app.core

import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable
import java.util.Locale

/**
 * A document attached to a tile (the file lives in Supabase Storage; this is its
 * row in the `attachments` table). Mirrors the iOS Attachment.
 */
@Serializable
data class Attachment(
    val id: String,
    @SerialName("tile_id") val tileId: String,
    val filename: String,
    @SerialName("storage_path") val storagePath: String,
    @SerialName("content_type") val contentType: String? = null,
    @SerialName("size_bytes") val sizeBytes: Long? = null,
    @SerialName("created_at") val createdAt: String,
) {
    /** Human-readable size, e.g. "1.2 MB"; null if unknown. */
    val displaySize: String? get() = sizeBytes?.let { formatBytes(it) }
}

private fun formatBytes(bytes: Long): String {
    if (bytes < 1024) return "$bytes B"
    val kb = bytes / 1024.0
    if (kb < 1024) return String.format(Locale.getDefault(), "%.0f KB", kb)
    return String.format(Locale.getDefault(), "%.1f MB", kb / 1024.0)
}
